File: app/src/main/java/maderski/bluetoothautoplaymusic/UI/TimePickerFragment.java
package maderski.bluetoothautoplaymusic.UI;

import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.StringDef;
import android.support.v4.app.DialogFragment;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.Gravity;
import android.widget.TextView;
import android.widget.TimePicker;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by Jason on 9/17/16.
 */
public class TimePickerFragment extends DialogFragment
        implements TimePickerDialog.OnTimeSetListener {
    private static final String TAG = "TimePickerFragment";

    @StringDef({
            TypeOfTimeSet.SCREEN_BRIGHTNESS_TIME,
            TypeOfTimeSet.MORNING_TIMESPAN,
            TypeOfTimeSet.EVENING_TIMESPAN
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface TypeOfTimeSet {
        String SCREEN_BRIGHTNESS_TIME = "screen_brightness_time";
        String MORNING_TIMESPAN = "morning_time_span";
        String EVENING_TIMESPAN = "evening_time_span";
    }

    private boolean mIsEndTime;
    private int mPreviouslySetTime;
    private String mPickerTitle;
    private String mTypeOfTimeSet;

    public TimePickerDialogListener dialogListener;

    public static TimePickerFragment newInstance(@TypeOfTimeSet String typeOfTimeSet, boolean isEndTime, int previouslySetTime, String title){
        Bundle args = new Bundle();
        args.putBoolean("picker_isDim", isEndTime);
        args.putInt("picker_previouslySetTime", previouslySetTime);
        args.putString("picker_title", title);
        args.putString("type_of_time_set", typeOfTimeSet);
        TimePickerFragment fragment = new TimePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        mIsEndTime = getArguments().getBoolean("picker_isDim");
        mPreviouslySetTime = getArguments().getInt("picker_previouslySetTime");
        mPickerTitle = getArguments().getString("picker_title");
        mTypeOfTimeSet = getArguments().getString("type_of_time_set");
        dialogListener = getActivity() instanceof TimePickerDialogListener ? (TimePickerDialogListener) getActivity() : null;

        //Create and set Title
        TextView tpfTitle = new TextView(getActivity());
        tpfTitle.setText(mPickerTitle);
        tpfTitle.setGravity(Gravity.CENTER_HORIZONTAL);

        //Get minutes
        int tempToGetMinutes = mPreviouslySetTime;
        while(tempToGetMinutes > 60){
            tempToGetMinutes -= 100;
        }
        int minute = tempToGetMinutes;
        //Get hour
        int hour = (mPreviouslySetTime - tempToGetMinutes)/100;

        // Create a new instance of TimePickerDialog
        TimePickerDialog timePickerDialog = new TimePickerDialog(getActivity(), this, hour, minute,
                DateFormat.is24HourFormat(getActivity()));
        //Add title to picker
        timePickerDialog.setCustomTitle(tpfTitle);
        //Return TimePickerDialog
        return timePickerDialog;
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        if(dialogListener != null) {
            dialogListener.onTimeCancel(mTypeOfTimeSet, mIsEndTime);
            Log.d(TAG, "Cancelled " + mTypeOfTimeSet + " time set...isEndTime: " + Boolean.toString(mIsEndTime));
        }
        super.onCancel(dialog);
    }

    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
        // Do something with the time chosen by the user
        int setTime = (hourOfDay * 100) + minute;
        Log.d(TAG, "Set time: " + Integer.toString(setTime));
        Log.d(TAG, "Is Dim: " + Boolean.toString(mIsEndTime));

        if(dialogListener != null)
            dialogListener.onTimeSet(mTypeOfTimeSet, mIsEndTime, view, hourOfDay, minute);


    }

    public interface TimePickerDialogListener {
        void onTimeSet(String typeOfTimeSet, boolean isEndTime, TimePicker view, int hourOfDay, int minute);
        void onTimeCancel(String typeOfTimeSet, boolean isEndTime);
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/ui/fragments/TimePickerFragment.kt
package maderski.bluetoothautoplaymusic.ui.fragments

import android.app.Dialog
import android.app.TimePickerDialog
import android.content.DialogInterface
import android.os.Bundle
import android.text.format.DateFormat
import android.util.Log
import android.view.Gravity
import android.widget.TextView
import android.widget.TimePicker
import androidx.annotation.StringDef
import androidx.fragment.app.DialogFragment

/**
 * Created by Jason on 9/17/16.
 */
class TimePickerFragment : DialogFragment(), TimePickerDialog.OnTimeSetListener {
    interface TimePickerDialogListener {
        fun onTimeSet(typeOfTimeSet: String, isEndTime: Boolean, view: TimePicker, hourOfDay: Int, minute: Int)
        fun onTimeCancel(typeOfTimeSet: String, isEndTime: Boolean)
    }

    private var isEndTime: Boolean = false
    private var previouslySetTime: Int = 0
    private var pickerTitle: String = ""
    private var typeOfTimeSet: String = ""

    private var dialogListener: TimePickerDialogListener? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        arguments?.let {
            isEndTime = it.getBoolean(PICKER_IS_DIM)
            previouslySetTime = it.getInt(PICKER_PREVIOUSLY_SET_TIME)
            pickerTitle = it.getString(PICKER_TITLE) ?: ""
            typeOfTimeSet = it.getString(TYPE_OF_TIME_SET) ?: ""
            dialogListener = if (requireActivity() is TimePickerDialogListener) requireActivity() as TimePickerDialogListener? else null
        }

        //Create and set Title
        val tpfTitle = TextView(requireActivity())
        tpfTitle.text = pickerTitle
        tpfTitle.gravity = Gravity.CENTER_HORIZONTAL

        //Get minutes
        var tempToGetMinutes = previouslySetTime
        while (tempToGetMinutes > 60) {
            tempToGetMinutes -= 100
        }
        val minute = tempToGetMinutes
        //Get hour
        val hour = (previouslySetTime - tempToGetMinutes) / 100

        // Create a new instance of TimePickerDialog
        val timePickerDialog = TimePickerDialog(activity, this, hour, minute,
                DateFormat.is24HourFormat(activity))
        //Add title to picker
        timePickerDialog.setCustomTitle(tpfTitle)
        //Return TimePickerDialog
        return timePickerDialog
    }

    override fun onCancel(dialog: DialogInterface) {
        if (dialogListener != null) {
            dialogListener?.onTimeCancel(typeOfTimeSet, isEndTime)
            Log.d(TAG, "Cancelled " + typeOfTimeSet + " time set...isEndTime: " + java.lang.Boolean.toString(isEndTime))
        }
        super.onCancel(dialog)

    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        // Do something with the time chosen by the user
        val setTime = hourOfDay * 100 + minute
        Log.d(TAG, "Set time: " + Integer.toString(setTime))
        Log.d(TAG, "Is Dim: " + java.lang.Boolean.toString(isEndTime))

        if (dialogListener != null)
            dialogListener?.onTimeSet(typeOfTimeSet, isEndTime, view, hourOfDay, minute)


    }

    companion object {
        private const val TAG = "TimePickerFragment"

        @StringDef(
                SCREEN_BRIGHTNESS_TIME,
                MORNING_TIMESPAN,
                EVENING_TIMESPAN,
                CUSTOM_TIMESPAN
        )
        @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
        annotation class TypeOfTimeSet

        const val SCREEN_BRIGHTNESS_TIME = "screen_brightness_time"
        const val MORNING_TIMESPAN = "morning_time_span"
        const val EVENING_TIMESPAN = "evening_time_span"
        const val CUSTOM_TIMESPAN = "custom_time_span"

        private const val PICKER_IS_DIM = "picker_isDim"
        private const val PICKER_PREVIOUSLY_SET_TIME = "picker_previouslySetTime"
        private const val PICKER_TITLE = "picker_title"
        private const val TYPE_OF_TIME_SET = "type_of_time_set"

        fun newInstance(@TypeOfTimeSet typeOfTimeSet: String, isEndTime: Boolean, previouslySetTime: Int, title: String): TimePickerFragment {
            val args = Bundle()
            args.putBoolean(PICKER_IS_DIM, isEndTime)
            args.putInt(PICKER_PREVIOUSLY_SET_TIME, previouslySetTime)
            args.putString(PICKER_TITLE, title)
            args.putString(TYPE_OF_TIME_SET, typeOfTimeSet)
            val fragment = TimePickerFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
